src/golang/lib/engine/databricks_executor.go

```diff
@@ -0,0 +1,132 @@
+	timeConfig *AqueductTimeConfig,
+	opExecMode operator.ExecutionMode,
+	databricksJobManager *job.DatabricksJobManager,
+) error {
+	inProgressOps := workflowRunMetadata.InProgressOps
+	completedOps := workflowRunMetadata.CompletedOps
+	dag := workflowDag
+
+	// Convert the operators into tasks
+	taskList, err := CreateTaskList(ctx, dag, databricksJobManager)
+	if err != nil {
+		return errors.Wrap(err, "Unable to convert operators to Databricks tasks.")
+	}
+
+	// Launch the workflow job with all tasks
+	_, err = databricksJobManager.LaunchMutlipleTaskJob(ctx, workflowName, taskList)
+	if err != nil {
+		return errors.Wrap(err, "Unable to launch workflow job on Databricks.")
+	}
+
+	for _, op := range dag.Operators() {
+		inProgressOps[op.ID()] = op
+	}
+
+	if len(inProgressOps) == 0 {
+		return errors.Newf("No initial operators to schedule.")
+	}
+
+	// Wait a little bit for all active operators to finish before exiting on failure.
+	defer waitForInProgressOperators(ctx, inProgressOps, timeConfig.OperatorPollInterval, timeConfig.CleanupTimeout)
+
+	start := time.Now()
+
+	for len(inProgressOps) > 0 {
+		if time.Since(start) > timeConfig.ExecTimeout {
+			return errors.New("Reached timeout waiting for workflow to complete.")
+		}
+
+		for _, op := range inProgressOps {
+			// Poll on the individual operator
+			execState, err := op.Poll(ctx)
+			if err != nil {
+				return err
+			}
+
+			if execState.Status == shared.PendingExecutionStatus {
+				continue
+			} else if execState.Status == shared.RunningExecutionStatus {
+				continue
+			}
+
+			if !execState.Terminated() {
+				return errors.Newf("Internal error: the operator is expected to have terminated, but instead has status %s", execState.Status)
+			}
+
+			// From here on we can assume that the operator has terminated.
+			if opExecMode == operator.Publish {
+				err = op.PersistResult(ctx)
+				if err != nil {
+					return errors.Wrapf(err, "Error when finishing execution of operator %s", op.Name())
+				}
+			}
+
+			// Add the operator to the completed stack, and remove it from the in-progress one.
+			if _, ok := completedOps[op.ID()]; ok {
+				return errors.Newf("Internal error: operator %s was completed twice.", op.Name())
+			}
+			completedOps[op.ID()] = op
+			delete(inProgressOps, op.ID())
+
+			time.Sleep(timeConfig.OperatorPollInterval)
+		}
+	}
+
+	if len(completedOps) != len(dag.Operators()) {
+		return errors.Newf("Internal error: %d operators were provided but only %d completed.", len(dag.Operators()), len(completedOps))
+	}
+	return nil
+}
+
+// Takes in a DAG and converts the operators into tasks within a Datbricks Job.
+func CreateTaskList(
+	ctx context.Context,
+	workflowDag dag_utils.WorkflowDag,
+	databricksJobManager *job.DatabricksJobManager,
+) ([]jobs.JobTaskSettings, error) {
+	dag := workflowDag
+	var taskList []jobs.JobTaskSettings
+
+	for _, op := range dag.Operators() {
+
+		// Get the upstream dependent operators
```

Comment:
Could you clarify how this prevents operators from running multiple times. For example, if i have a simple dag with operators: A --> B --> C, my understanding is that this function will create a task for each of those operators with the appropriate parent dependencies:
- A has none
- B has A
- C has B

But then isn't the `databricksJobManager.LaunchMutlipleTaskJob` call in the function above running all of these tasks as part of one job?

Reply:
`CreateTaskList` doesn't actually launch any jobs. Here we are just assembling that task dependency graph that databricks needs. The translation works like this:
Workflow -> DatabricksJob
Operator -> Task 

A task is considered a "step" within a job. When the dependent tasks are complete Databricks' own scheduler knows to start the next task.

Within each task specification we have to provide the parent tasks (Databricks has no notion of input/output artifacts so this is how they create a dependency graph.) `CreateTaskList` creates a list of all these task specifications.